File: Subsets/Rust/src/solution_4.rs
// https://www.lintcode.com/problem/17/
// https://leetcode.cn/problems/subsets/submissions/

pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
    let mut nums = nums;

    if nums.is_empty() {
        return Vec::new();
    }

    let mut queue: Vec<Vec<i32>> = vec![vec![]];
    nums.sort();

    let mut index = 0;

    while index < queue.len() {
        let subset = queue[index].clone();
        index += 1;

        for &num in &nums {
            if !subset.is_empty() && num <= *subset.last().unwrap() {
                continue;
            }

            let mut new_subsets = subset.clone();
            new_subsets.push(num);
            queue.push(new_subsets);
        }
    }

    queue
}

File: Unique Paths/Javascript/solution_4.js
// https://leetcode.cn/problems/unique-paths/
// https://www.lintcode.com/problem/114/

// 自底向上进行 DP

const uniquePaths = function (m, n) {
  let dp = [];

  for (let i = 0; i < m; i++) {
    dp.push(Array(n).fill(0));
  }

  for (let i = 0; i < n; i++) {
    dp[m - 1][i] = 1;
  }

  for (let i = 0; i < m; i++) {
    dp[i][n - 1] = 1;
  }

  for (let i = m - 2; i >= 0; i--) {
    for (let j = n - 2; j >= 0; j--) {
      dp[i][j] = dp[i][j + 1] + dp[i + 1][j];
    }
  }

  return dp[0][0];
};

console.log(uniquePaths(3, 7));
